RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,468 @@
+- Rudimentary monitoring of total resource usage will be available through the the CLI if needed.
+
+## Detailed Design
+
+### 1. Library
+
+#### a. In-memory Job Logger
+- Implement a thread-safe circular buffer to store logs with a configurable maximum size
+- Use a read-write mutex to ensure thread-safety for concurrent read/write operations
+- Implement a pub/sub pattern for real-time log streaming:
+  - Use channels for each subscriber
+  - Implement a fan-out pattern to broadcast new log entries to all subscribers
+- Provide methods for adding logs, retrieving all logs, and subscribing to log streams
+- Implement log rotation to handle long-running jobs and prevent memory exhaustion
+
+```go
+// Logger represents an in-memory circular buffer for storing job logs.
+type Logger struct {
+	mu          sync.RWMutex
+	logs        []string
+	maxSize     int
+	subscribers map[chan string]struct{}
+}
+
+// InitLogger creates a new Logger with the specified maximum size.
+func InitLogger(maxSize int) *Logger
+
+// Log adds a new log entry to the logger.
+func (l *Logger) Log(log string)
+
+// Retrieves all logs currently stored in the logger.
+func (l *Logger) GetLogs() []string
+
+// Returns a channel that receives new log entries.
+func (l *Logger) Subscribe() <-chan string
+
+// Removes a subscriber from the logger.
+func (l *Logger) Unsubscribe(ch <-chan string)
+
+// Closes all subscriber channels and clears the subscriber list.
+func (l *Logger) Close()
+```
+
+#### b. Job Executor
+- Use `os/exec` package to create and manage child processes
+- Implement non-blocking I/O for stdout/stderr capture using `io.Pipe()`
+- Use a separate goroutine for each of stdout and stderr to prevent blocking
+- Implement a context-based cancellation mechanism for graceful job termination
+- Monitor resource usage using cgroups
+- Provide callbacks for resource usage updates
+- Use a Go channel to periodically read current resource usage (memory, I/O, and CPU usage) from the associated cgroup
+
+```go
+// Executor is responsible for running jobs and managing their lifecycle.
+type Executor struct {
+	cmd     *exec.Cmd
+	logger  *logger.Logger
+    // Context to handle propagate state and cancellations
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started time.Time
+	stopped time.Time
+	mu      sync.Mutex
+}
+
+// NewExecutor creates a new Executor for the given command and arguments.
+func NewExecutor(command string, args []string, logger *Logger) *Executor
+
+// Start begins the execution of the job.
+func (e *Executor) Start() error
+
+// Stop terminates the running job.
+func (e *Executor) Stop() error
+
+// Wait blocks until the job is completed and returns its exit status.
+func (e *Executor) Wait() error
+```
+
+#### c. Job Manager
+- Manage job lifecycle (pending, running, completed)
+- Maintain job queue and schedule based on resource availability
```

Comment:
Scheduling is not a challenge requirement

Reply:
thanks for pointing that out! I got rid of it.